File: hostel_management_system_backend/Controllers/HostelReviewsController.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using hostel_management_system_backend.Exceptions;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("api/hostels/{hostelId:guid}/reviews")]
public sealed class HostelReviewsController : ControllerBase
{
    private readonly IHostelReviewsService _service;

    public HostelReviewsController(IHostelReviewsService service)
    {
        _service = service;
    }

    [AllowAnonymous]
    [HttpGet]
    public async Task<ActionResult<List<HostelReviewReadDto>>> GetForHostel(Guid hostelId, CancellationToken cancellationToken)
    {
        var reviews = await _service.GetForHostelAsync(hostelId, cancellationToken);
        return Ok(reviews);
    }

    [AllowAnonymous]
    [HttpGet("summary")]
    public async Task<ActionResult<HostelRatingSummaryDto>> GetSummary(Guid hostelId, CancellationToken cancellationToken)
    {
        var summary = await _service.GetSummaryAsync(hostelId, cancellationToken);
        return Ok(summary);
    }

    [Authorize]
    [HttpPost]
    public async Task<ActionResult<HostelReviewReadDto>> Create(Guid hostelId, [FromBody] HostelReviewCreateDto dto, CancellationToken cancellationToken)
    {
        var userId = GetUserIdOrThrow();
        var created = await _service.CreateAsync(hostelId, userId, dto, cancellationToken);
        return CreatedAtAction(nameof(GetForHostel), new { hostelId }, created);
    }

    [Authorize]
    [HttpPut("{reviewId:guid}")]
    public async Task<ActionResult<HostelReviewReadDto>> Update(Guid hostelId, Guid reviewId, [FromBody] HostelReviewUpdateDto dto, CancellationToken cancellationToken)
    {
        var userId = GetUserIdOrThrow();
        var isAdmin = User.IsInRole(UserRole.Admin.ToString());
        var updated = await _service.UpdateAsync(hostelId, reviewId, userId, isAdmin, dto, cancellationToken);
        return Ok(updated);
    }

    [Authorize]
    [HttpDelete("{reviewId:guid}")]
    public async Task<IActionResult> Delete(Guid hostelId, Guid reviewId, CancellationToken cancellationToken)
    {
        var userId = GetUserIdOrThrow();
        var isAdmin = User.IsInRole(UserRole.Admin.ToString());
        await _service.DeleteAsync(hostelId, reviewId, userId, isAdmin, cancellationToken);
        return NoContent();
    }

    private Guid GetUserIdOrThrow()
    {
        var sub = User.FindFirstValue(JwtRegisteredClaimNames.Sub);
        if (!Guid.TryParse(sub, out var userId))
        {
            throw new UnauthorizedException("Invalid access token.", "invalid_token");
        }

        return userId;
    }
}

File: src/hostels/reviews/hostel-reviews.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
  UseGuards,
} from '@nestjs/common';

import { CurrentUserDecorator } from '../../common/decorators/current-user.decorator';
import { UserRole } from '../../common/enums/app.enums';
import { JwtAuthGuard } from '../../common/guards/jwt-auth.guard';
import { CurrentUser } from '../../common/interfaces/current-user.interface';
import {
  HostelRatingSummaryDto,
  HostelReviewCreateDto,
  HostelReviewReadDto,
  HostelReviewUpdateDto,
} from './dto/hostel-reviews.dto';
import { HostelReviewsService } from './hostel-reviews.service';

@Controller('hostels/:hostelId/reviews')
export class HostelReviewsController {
  constructor(private readonly hostelReviewsService: HostelReviewsService) {}

  @Get()
  getForHostel(@Param('hostelId', ParseUUIDPipe) hostelId: string): Promise<HostelReviewReadDto[]> {
    return this.hostelReviewsService.getForHostel(hostelId);
  }

  @Get('summary')
  getSummary(@Param('hostelId', ParseUUIDPipe) hostelId: string): Promise<HostelRatingSummaryDto> {
    return this.hostelReviewsService.getSummary(hostelId);
  }

  @Post()
  @UseGuards(JwtAuthGuard)
  create(
    @Param('hostelId', ParseUUIDPipe) hostelId: string,
    @CurrentUserDecorator() currentUser: CurrentUser,
    @Body() dto: HostelReviewCreateDto,
  ): Promise<HostelReviewReadDto> {
    return this.hostelReviewsService.create(hostelId, currentUser.userId, dto);
  }

  @Put(':reviewId')
  @UseGuards(JwtAuthGuard)
  update(
    @Param('hostelId', ParseUUIDPipe) hostelId: string,
    @Param('reviewId', ParseUUIDPipe) reviewId: string,
    @CurrentUserDecorator() currentUser: CurrentUser,
    @Body() dto: HostelReviewUpdateDto,
  ): Promise<HostelReviewReadDto> {
    return this.hostelReviewsService.update(
      hostelId,
      reviewId,
      currentUser.userId,
      currentUser.role === UserRole.Admin,
      dto,
    );
  }

  @Delete(':reviewId')
  @UseGuards(JwtAuthGuard)
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(
    @Param('hostelId', ParseUUIDPipe) hostelId: string,
    @Param('reviewId', ParseUUIDPipe) reviewId: string,
    @CurrentUserDecorator() currentUser: CurrentUser,
  ) {
    await this.hostelReviewsService.delete(
      hostelId,
      reviewId,
      currentUser.userId,
      currentUser.role === UserRole.Admin,
    );
  }
}
